packages/cozy-konnector-libs/libs/signin.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * The goal of this function is to provide an handy method to log the user in,
+ * on html form pages. On success, it resolves a promise with a parsed body.
+ *
+ * Errors:
+ *
+ * - LOGIN_FAILED if the validate predicate is false
+ * - INVALID_FORM if the element matched by `formSelector` is not a form or has
+ *   no `action` attribute
+ * - UNKNOWN_PARSING_STRATEGY if `parseStrategy` is not one of the accepted
+ *   values : `raw`, `cheerio`, `json`.
+ * - VENDOR_DOWN if a request throws a RequestError
+ *
+ * It does not submit values provided through `select` tags, except if populated
+ * by user with `formData`.
+ * Limit between `baseUrl` and `page` is important. The form `action` will be
+ * concatenated the same way `page` is, to constitute the submission url. For
+ * now, a form with an absolute `action` is not comptabile with this function.
+ *
+ * - `pageUrl` is the url to access the html form
+ *
+ * - `formSelector` is used by cheerio to uniquely identify the form in which to
+ *   log in
+ *
+ * - `formData` is an object of { name: value, … }. It is used to populate the
+ *   form, in the proper inputs with the same name as the properties of this
+ *   object, before submitting it.
+ *
+ * - `parseStrategy` allow the user to resolve `signin` with a preparsed body.
+ *   The choice of the strategy for the parsing is one of : `raw`, `json` or
+ *   `cheerio`. `cheerio` being the default.
+ *
+ * - `validate` is a predicate takin two arguments `statusCode` and
+ *   `parsedBody`. If it is false, `LOGIN_FAILED` is thrown, otherwise the
+ *   konnector continues.
+ *
+ * - `opts` allows to pass eventual options to the `signin`'s `requestFactory`.
+ *   It could be useful for pages using `latin1` `encoding` for instance.
+ *
+ * @module signin
+ */
+const errors = require('../helpers/errors')
+const rerrors = require('request-promise/errors')
+const log = require('cozy-logger').namespace('cozy-konnector-libs')
+const requestFactory = require('./request')
+const url = require('url')
+
+module.exports = function signin (
+  pageUrl,
+  formSelector,
+  formData,
+  parseStrategy = 'cheerio',
+  validate = defaultValidate,
+  opts = {}) {
+  const defaultOpts = { jar: true, cheerio: true, json: false }
+
+  const rq = requestFactory({
+    ...opts,
+    ...defaultOpts
```

Comment:
`...opts` and `...defaultOts` should be switched as here it is not possible to overwrite a default option.

Reply:
It is made on purpose, as defaultOpts contains `cheerio: true` and that this behavior is needed for core implementation of `signin`.
Everything that is un-overwritteable by `defaultOpts` can be customized for the output.

If you want output as `json` you switch your parsing strategy to `json`.

I agree on one thing though, if you want to provide a specific `jar` you won't be able to do it.

I suggest the following:

```js
const rq = requestFactory({
  jar: true,
  ...opts,
  cheerio: true,
  json: false
})
```

WDYT?

Reply:
I commit this then ;-)